SpatialGDK/Source/SpatialGDK/Private/Utils/EntityPool.cpp

```diff
@@ -0,0 +1,166 @@
+		{
+			return !Element.bExpired;
+		});
+
+		EntityRange NewEntityRange = {};
+		NewEntityRange.CurrentEntityId = Op.first_entity_id;
+		NewEntityRange.LastEntityId = Op.first_entity_id + (Op.number_of_entity_ids - 1);
+		NewEntityRange.EntityRangeId = NextEntityRangeId++;
+
+		UE_LOG(LogSpatialEntityPool, Log, TEXT("Reserved %d entities, caching in pool, Entity IDs: (%d, %d) Range ID: %d"), Op.number_of_entity_ids, Op.first_entity_id, NewEntityRange.LastEntityId, NewEntityRange.EntityRangeId);
+
+		ReservedEntityIDRanges.Add(NewEntityRange);
+
+		TWeakObjectPtr<UEntityPool> WeakEntityPoolPtr = this;
+		FTimerHandle ExpirationTimer;
+		TimerManager->SetTimer(ExpirationTimer, [WeakEntityPoolPtr, ExpiringEntityRangeId = NewEntityRange.EntityRangeId]()
+		{
+			if (!WeakEntityPoolPtr.IsValid())
+			{
+				return;
+			}
+
+			WeakEntityPoolPtr->OnEntityRangeExpired(ExpiringEntityRangeId);
+		}, 180.0f, false); // 3 minute timeout for reserved Entity IDs, timeout for Runtime is 5 minutes
+
+		bIsReady = true;
+		bIsAwaitingResponse = false;
+	});
+
+	// Reserve the Entity IDs
+	Worker_RequestId ReserveRequestID = NetDriver->Connection->SendReserveEntityIdsRequest(EntitiesToReserve);
+	bIsAwaitingResponse = true;
+
+	// Add the spawn delegate
+	Receiver->AddReserveEntityIdsDelegate(ReserveRequestID, CacheEntityIDsDelegate);
+}
+
+bool UEntityPool::IsReady()
+{
+	return bIsReady;
+}
+
+void UEntityPool::OnEntityRangeExpired(uint32 ExpiringEntityRangeId)
+{
+	UE_LOG(LogSpatialEntityPool, Log, TEXT("Entity range expired! Range ID: %d"), ExpiringEntityRangeId);
+
+	int32 FoundEntityRangeIndex = ReservedEntityIDRanges.IndexOfByPredicate([ExpiringEntityRangeId](const EntityRange& Element)
+	{
+		return Element.EntityRangeId == ExpiringEntityRangeId;
+	});
+
+	if (FoundEntityRangeIndex == INDEX_NONE)
+	{
+		// This entity range has already been cleaned up as a result of running out of Entity IDs.
+		UE_LOG(LogSpatialEntityPool, Log, TEXT("Entity range ID: %d has already been depleted"), ExpiringEntityRangeId);
+		return;
+	}
+
+	if (FoundEntityRangeIndex < ReservedEntityIDRanges.Num() - 1)
+	{
+		// This is not the most recent entity range, just clean up without requesting additional IDs.
+		UE_LOG(LogSpatialEntityPool, Log, TEXT("Newer range detected, cleaning up Entity range ID: %d without new request"), ExpiringEntityRangeId);
+		ReservedEntityIDRanges.RemoveAt(FoundEntityRangeIndex);
+	}
+	else
+	{
+		// Reserve then cleanup
+		if (!bIsAwaitingResponse)
+		{
+			UE_LOG(LogSpatialEntityPool, Log, TEXT("Reserving new Entity range to replace Entity range ID: %d"), ExpiringEntityRangeId);
+			ReserveEntityIDs(GetDefault<USpatialGDKSettings>()->EntityPoolRefreshCount);
+		}
+		// Mark this entity range as expired, so it gets cleaned up when we receive a new entity range from Spatial.
+		ReservedEntityIDRanges[FoundEntityRangeIndex].bExpired = true;
+	}
+}
+
+Worker_EntityId UEntityPool::Pop()
```

Comment:
Could this function be renamed into something more descriptive?

Reply:
Changed to `GetNextEntityId`